instrumentation/github.com/99designs/gqlgen/otelgqlgen/gqlgen.go

```diff
@@ -0,0 +1,162 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelgqlgen
+
+import (
+	"context"
+	"os"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+
+	otelcontrib "go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerName         = "go.opentelemetry.io/contrib/instrumentation/github.com/99designs/gqlgen/otelgqlgen"
+	extensionName      = "OpenTelemetry"
+	complexityLimit    = "ComplexityLimit"
+	defaultServiceName = "GraphQLService"
+)
+
+type Tracer struct {
+	serviceName             string
+	complexityExtensionName string
+	tracer                  oteltrace.Tracer
+}
+
+var _ interface {
+	graphql.HandlerExtension
+	graphql.ResponseInterceptor
+	graphql.FieldInterceptor
+} = Tracer{}
+
+func (a Tracer) ExtensionName() string {
+	return extensionName
+}
+
+func (a Tracer) Validate(_ graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	ctx, span := a.tracer.Start(ctx, operationName(ctx), oteltrace.WithSpanKind(oteltrace.SpanKindServer))
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	oc := graphql.GetOperationContext(ctx)
+
+	serviceName := os.Getenv("OTEL_SERVICE_NAME")
+	if serviceName == "" {
+		serviceName = defaultServiceName
+	}
+	span.SetAttributes(
+		semconv.ServiceNameKey.String(serviceName),
+		RequestQuery(oc.RawQuery),
+	)
+	complexityExtension := a.complexityExtensionName
+	if complexityExtension == "" {
+		complexityExtension = complexityLimit
+	}
+	complexityStats, ok := oc.Stats.GetExtension(complexityExtension).(*extension.ComplexityStats)
+	if !ok {
+		// complexity extension is not used
+		complexityStats = &extension.ComplexityStats{}
+	}
+
+	if complexityStats.ComplexityLimit > 0 {
+		span.SetAttributes(
+			RequestComplexityLimit(int64(complexityStats.ComplexityLimit)),
+			RequestOperationComplexity(int64(complexityStats.Complexity)),
+		)
+	}
+
+	span.SetAttributes(RequestVariables(oc.Variables)...)
+
+	return next(ctx)
```

Comment:
shouldn't we do something more or less like this?

```go
resp := nex(ctx)
if len(resp.Errors) > 0 {
  span.RecordError(err)
  span.SetStatus(codes.Error, err.Error())
}

Reply:
same on this. this will be correct?

```
	resp := next(ctx)
	if len(resp.Errors) > 0 {
		span.RecordError(fmt.Errorf(resp.Errors.Error()))
		span.SetStatus(codes.Error, resp.Errors.Error())
		span.SetAttributes(ResolverErrors(resp.Errors)...)
	}

	return resp
```

Reply:
Looks OK. And a unit test if possible 😉